src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/RuntimeAssemblyBuilder.cs

```diff
@@ -17,13 +17,62 @@
 
 namespace System.Reflection.Emit
 {
-    public sealed partial class AssemblyBuilder : Assembly
+    public partial class AssemblyBuilder
+    {
+        [RequiresDynamicCode("Defining a dynamic assembly requires dynamic code.")]
+        [DynamicSecurityMethod] // Required to make Assembly.GetCallingAssembly reliable.
+        public static AssemblyBuilder DefineDynamicAssembly(AssemblyName name, AssemblyBuilderAccess access)
+            => DefineDynamicAssembly(name, access, null, Assembly.GetCallingAssembly());
+
+        [RequiresDynamicCode("Defining a dynamic assembly requires dynamic code.")]
+        [DynamicSecurityMethod] // Required to make Assembly.GetCallingAssembly reliable.
+        public static AssemblyBuilder DefineDynamicAssembly(
+            AssemblyName name,
+            AssemblyBuilderAccess access,
+            IEnumerable<CustomAttributeBuilder>? assemblyAttributes)
+                => DefineDynamicAssembly(name, access, assemblyAttributes, Assembly.GetCallingAssembly());
+
+        private static AssemblyBuilder DefineDynamicAssembly(
+            AssemblyName name,
+            AssemblyBuilderAccess access,
+            IEnumerable<CustomAttributeBuilder>? assemblyAttributes,
+            Assembly? callingAssembly)
+        {
+            ArgumentNullException.ThrowIfNull(name);
+
+            if (access != AssemblyBuilderAccess.Run && access != AssemblyBuilderAccess.RunAndCollect)
+            {
+                throw new ArgumentException(SR.Format(SR.Arg_EnumIllegalVal, (int)access), nameof(access));
+            }
+
+            if (callingAssembly == null)
+            {
+                // Called either from interop or async delegate invocation. Rejecting because we don't
+                // know how to set the correct context of the new dynamic assembly.
+                throw new InvalidOperationException();
+            }
+
+            EnsureDynamicCodeSupported();
+
+            AssemblyLoadContext? assemblyLoadContext =
+                AssemblyLoadContext.CurrentContextualReflectionContext ?? AssemblyLoadContext.GetLoadContext(callingAssembly);
+
+            if (assemblyLoadContext == null)
+            {
+                throw new InvalidOperationException();
```

Comment:
Is this ever expected?

Reply:
Based on the nullable annotation of this APIs they are both nullable, though not sure when, especially for the AssemblyLoadContext.CurrentContextualReflectionContext, maybe try adding test for this and the above case later